docker-entrypoint.sh

```diff
@@ -31,30 +36,33 @@ drop_privs_cmd() {
     fi
 }
 
+inject_config_overrides() {
+    # Find and replace existing config
+    sed -i -e "s/jobmanager.rpc.address: localhost/jobmanager.rpc.address: ${JOB_MANAGER_RPC_ADDRESS}/g" "$FLINK_HOME/conf/flink-conf.yaml"
+    sed -i -e "s/jobmanager.rpc.port: 6123/jobmanager.rpc.port: ${JOB_MANAGER_RPC_PORT}/g" "$FLINK_HOME/conf/flink-conf.yaml"
+    sed -i -e "s/rest.port: 8081/rest.port: ${JOB_MANAGER_WEB_PORT}/g" "$FLINK_HOME/conf/flink-conf.yaml"
+    sed -i -e "s/taskmanager.numberOfTaskSlots: 1/taskmanager.numberOfTaskSlots: $TASK_MANAGER_NUMBER_OF_TASK_SLOTS/g" "$FLINK_HOME/conf/flink-conf.yaml"
+    # Inject new previously unspecified config
+    echo "blob.server.port: ${BLOB_SERVER_PORT}" >> "$FLINK_HOME/conf/flink-conf.yaml"
+    echo "query.server.port: ${QUERY_SERVER_PORT}" >> "$FLINK_HOME/conf/flink-conf.yaml"
+
+    echo "config file: " && grep '^[^\n#]' "$FLINK_HOME/conf/flink-conf.yaml"
+}
+
 if [ "$1" = "help" ]; then
     echo "Usage: $(basename "$0") (jobmanager|taskmanager|local|help)"
     exit 0
 elif [ "$1" = "jobmanager" ]; then
     echo "Starting Job Manager"
-    sed -i -e "s/jobmanager.rpc.address: localhost/jobmanager.rpc.address: ${JOB_MANAGER_RPC_ADDRESS}/g" "$FLINK_HOME/conf/flink-conf.yaml"
-    echo "blob.server.port: 6124" >> "$FLINK_HOME/conf/flink-conf.yaml"
-    echo "query.server.port: 6125" >> "$FLINK_HOME/conf/flink-conf.yaml"
-
-    echo "config file: " && grep '^[^\n#]' "$FLINK_HOME/conf/flink-conf.yaml"
+    inject_config_overrides
     exec $(drop_privs_cmd) flink "$FLINK_HOME/bin/jobmanager.sh" start-foreground cluster
 elif [ "$1" = "taskmanager" ]; then
-    TASK_MANAGER_NUMBER_OF_TASK_SLOTS=${TASK_MANAGER_NUMBER_OF_TASK_SLOTS:-$(grep -c ^processor /proc/cpuinfo)}
-
-    sed -i -e "s/jobmanager.rpc.address: localhost/jobmanager.rpc.address: ${JOB_MANAGER_RPC_ADDRESS}/g" "$FLINK_HOME/conf/flink-conf.yaml"
-    sed -i -e "s/taskmanager.numberOfTaskSlots: 1/taskmanager.numberOfTaskSlots: $TASK_MANAGER_NUMBER_OF_TASK_SLOTS/g" "$FLINK_HOME/conf/flink-conf.yaml"
-    echo "blob.server.port: 6124" >> "$FLINK_HOME/conf/flink-conf.yaml"
-    echo "query.server.port: 6125" >> "$FLINK_HOME/conf/flink-conf.yaml"
-
-    echo "Starting Task Manager"
-    echo "config file: " && grep '^[^\n#]' "$FLINK_HOME/conf/flink-conf.yaml"
+    inject_config_overrides
     exec $(drop_privs_cmd) flink "$FLINK_HOME/bin/taskmanager.sh" start-foreground
 elif [ "$1" = "local" ]; then
     echo "Starting local cluster"
+    inject_config_overrides
+    $(drop_privs_cmd) flink "$FLINK_HOME/bin/taskmanager.sh" start
```

Comment:
I'm aware that local mode is broken now in 1.5, but I think it's against Docker best practices to run an orphaned process alongside the PID 1.

I think we either need to update the docs to suggest users run both a jobmanager and taskmanager themselves, linking them, or introduce a process supervisor. Even though it keeps things "broken", could you remove this line for now, pending a real fix?

Reply:
Hey Patrick, thanks for the comments.

How would you feel about me using `supervisord` to manage the job and task managers? The container will have little use for me (automated testing) if I can't run it as a local cluster, so I'm hesitant to lose that bit (though you are completely right about me violating docker good practices in the current form!).

Reply:
I'm a fan of supervisord, but in this case I'm afraid it will grow the image unnecessarily since it depends on python, and I think we should optimize for the typical case which doesn't require it.

Daemontools is another possibility, but I admittedly haven't used it in many years—maybe there's somethings else preferred by the Docker community these days? These "official images" are meant to set the standard for following Docker best practices, so I think I'm unwilling to push this as-is.

I've also posed the question to some Flink devs, and someone suggested we could expose MiniCluster from the CLI scripts in Flink itself to provide this behavior once again.